### ISSUE TITLE
revisiting an open file sometimes stalls at "Loading..."; directory listing doesn't update

### ISSUE THREAD

**Issue by DrXyzzy** [sagemathinc/cocalc#4630]
I am seeing this several times a week, with Chrome and Firefox on macOS 10.15.5. If/when I find more of a pattern, I'll update this ticket. Most recently seen as of this writing: June 2, 2020.

1. Open a markdown file in some project; call it A.
1. Open another file in project A. Don't close the tab for the first file. 
1. Open another project; call it B. Don't close the tab for project A. Work in B for an hour or more.
1. Return to project A and open the markdown file from step 1. About 1 time in 50, the "Loading ..." message and spinner appear and stay there apparently indefinitely - over a minute at least. The javascript console does not show any error messages. The network connection icon is grey. Refreshing the browser always causes the file to be displayed immediately.

I see this with other file types (.j, .rst), but am not sure it can happen with every file type.

**Comment by DrXyzzy** [comment]
I just saw it again, this time with Firefox, opening a .sagews file. The project shows as running and the network icon is normal (gray). Web console log shows no recent errors.

**Comment by DrXyzzy** [comment]
More clues. I just saw the problem open stalls at "Loading..." again with a .md file in Chrome and noticed:

- I am unable to open other already-open files in the same project of the following types: .md, .txt, .sage-chat, .tasks; opening them also stalls
- I am **able** to open files of type .term and .pdf. But when I tab to one of the other file types, they are still stalled at "Loading..."
- the files don't have to be open in the project when this happens. I can click any file in the log, and if the file type is one of those that stalls, opening this file also stalls. 
- closing and re-opening a stalled file does not fix the problem
- viewing the Files list shows the root directory, but viewing subdirectories also stalls
- doing a hub reconnect does not change the problem
- I can switch to other projects and open any file, then come back to the problem project and the problem is still there, as long as I don't refresh the browser
- while viewing the problem in Chrome, I logged into the same account from Firefox and opened the same project; at this time there were no already-open files in the Firefox session; as soon as the root Files listing appeared in Firefox, the file tabs in Chrome became unstuck, just as if I had refreshed Chrome

**Comment by williamstein** [comment]
When exactly did you restart the project where you are observing this?   Also, did you refresh your browser recently?

**Comment by DrXyzzy** [comment]
The project in the second comment was started about 2.5 days ago. I'm not sure, but think browser (Firefox) I saw the problem in was refreshed or started a few hours before that.

The project I reported on in the third comment has been running for 41 days.  The browser (Chrome) had been refreshed maybe 25 minutes before I noticed the problem.

**Comment by williamstein** [comment]
Restart both of your projects, then continue testing.  There have been recent updates to the project image that could have fixed this bug.

**Comment by DrXyzzy** [comment]
ok

**Comment by williamstein** [comment]
And I've been working pretty intensely on cocalc for two days -- but with restarted projects -- and haven't seen this problem at all.  So I might have fixed it a few days ago.

**Comment by DrXyzzy** [comment]
Since the previous comments, I've seen the problem twice on projects and browsers restarted recently (i.e. after the above comments). Behavior may have changed. Below are two new observations.

In this latest instance, seen with Chrome, the browser was restarted this morning, and the project settings page shows `project started 1 day 5 hours 48 minutes ago`.

- There is one .rst file open that I **can** visit that displays correctly. All other .rst files I try to open stall, including those that have tabs showing in the session and others that I select from project log and Files list.
- Google Developer Javascript Console shows many notices like the following. There are over two hundred now, about 20 minutes after first seeing the problem. These messages are arriving irregularly every 1-10 seconds.

    ```
    smc-20e59b60ae1b97c39a70.cacheme.js?20e59b60ae1b97c39a70:395 [Violation] 'setInterval' handler took 140ms
    ```

**Comment by williamstein** [comment]
> Google Developer Javascript Console shows many notices like the following. 

I'm pretty sure the console has shown those for many years.

Interesting that you still have the problem.  I've been working all day every day on cocalc this week and haven't seen this problem this week.  I wonder what is different?    I'm not using rst files and I'm only using Chrome.

**Comment by haraldschilly** [comment]
FWIW, I just had this "stuck at Loading..." as well. For my dev project, 3 days running. I'll restart it and see if that helps. Ok, restart helped, didn't refresh browser. Next time this happens I'll try to just close and open the editor tab.

side note: I was using `open` in a terminal.

**Comment by williamstein** [comment]
I just saw this in a cc-in-cc dev project, and the relevant log is:
```
~$ grep sync:fda6bb707dc3527c4dc040d2273b624dec518f79 log
2020-06-14T04:44:56.570Z - debug: synctable_channel {"listings":[{"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","path":null,"listing":null,"time":null,"interest":null,"missing":null,"error":null,"deleted":null}]} sync:fda6bb707dc3527c4dc040d2273b624dec518f79
2020-06-14T04:44:56.571Z - debug: SyncTableChannel('sync:fda6bb707dc3527c4dc040d2273b624dec518f79', '{"listings":[{"deleted":null,"error":null,"interest":null,"listing":null,"missing":null,"path":null,"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","time":null}]}'):  creating new sync channel (persistent=false, ephemeral=false)
2020-06-14T04:44:56.571Z - debug: SyncTableChannel('sync:fda6bb707dc3527c4dc040d2273b624dec518f79', '{"listings":[{"deleted":null,"error":null,"interest":null,"listing":null,"missing":null,"path":null,"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","time":null}]}'):  init_handlers
2020-06-14T04:44:56.572Z - debug: SyncTableChannel('sync:fda6bb707dc3527c4dc040d2273b624dec518f79', '{"listings":[{"deleted":null,"error":null,"interest":null,"listing":null,"missing":null,"path":null,"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","time":null}]}'):  init_synctable
2020-06-14T04:44:56.572Z - debug: SyncTableChannel('sync:fda6bb707dc3527c4dc040d2273b624dec518f79', '{"listings":[{"deleted":null,"error":null,"interest":null,"listing":null,"missing":null,"path":null,"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","time":null}]}'):  init_synctable -- persistent (but no changefeeds)
2020-06-14T04:44:56.573Z - debug: SyncTableChannel('sync:fda6bb707dc3527c4dc040d2273b624dec518f79', '{"listings":[{"deleted":null,"error":null,"interest":null,"listing":null,"missing":null,"path":null,"project_id":"aff5e9e4-2921-4af5-a9d8-15ed5a14cd0f","time":null}]}'):  created synctable -- waiting for connected state
```

Note that it ends with:  "created synctable -- waiting for connected state".  This means the project is waiting to get a connection from some hub before finishing initialization of the listings table, as it should.  So the real root problem is that there's no connection from any hub.  The log has an entry like this when there's a connection from a hub (this is from a different project):
```
~$ grep "hello from hub" log
2020-06-14T04:45:59.542Z - debug: hello from hub -- sending back our version = 1592094807
```
In any case, this helps me a lot in fixing this.  I'm going to prioritize finding a **new way** to ensure that any "in use" project always has an active connection from some hub.   One idea is to have the project send a message to a browser client saying "heh, I need a hub", and the brower client would then send a message to its hub saying "connect to this project please".   That all fits with our security  model and should work exactly when the project is actually in use by a client.

**Comment by williamstein** [comment]
I just observed this and got some new information.

1. I opened my personal "dev1" project, after not using it for two days.
2. I browsed around and when I hit a directory I hadn't opened in a long time, the listing was taking forever to be updated.
3. I checked with the project's log and it said it had got the request to update the listing, and it had done it.
4. I looked in the database, and the directory listing was updated there.
5. I opened a different browser tab on the same directory and it immediately showed the listing.   Moreover, my first browser tab, which wasn't updating, immediately updated upon doing this.  Also, I could browse around at many new directories and everything was fine.  Basically, just opening a new browser tab completely eliminated the problem with the old one, without having to refresh the old one or do anything.

This is of course exactly what @DrXyzzy reported above.  When opening a file one of the steps is I think consulting the directory listing (to know whether to open an existing file or create a new one), and that's where things are hanging.    Also, the fact that the listing was being computed and placed in the database means that there is a connection from some hub to the project this whole time.    Being offline for a while, then connecting again is probably also very relevant.  Hopefully this will help me fix this bug.

**Comment by williamstein** [comment]
OK, I saw this again with the same project after a browser suspend/resume.  This time the little file listing spinners are staying visible:

![image](https://user-images.githubusercontent.com/1276278/85186491-2a97b500-b24e-11ea-96ac-d507e7d81ba2.png)

Opening the same directories in another browser tab works fine, but the original browser tab is NOT suddenly working.  Closing and opening the project tab in the original browser tab did fix the problem.

**Comment by williamstein** [comment]
Having looked closely at what happens when this failed on me in a cc-in-cc dev project, there it was 100% definitely due to the hub not being connected to the project, and that very much seemed to be due to some crappy callback version of reuseinflight in some old code... that is not even run on cocalc.com.  In any case, I think I fixed that code in the commit 94d2b09.

This definitely makes me think this is in general an old bug in smc-hub, and perhaps it isn't new at all.  Maybe it's just that so many other things are very stable in cocalc that this bug finally gets noticed.  (It also might be that the bug only became visible when we simplified the architecture of the hub services in kucalc by eliminating hub-user-api and a few other microservices.)  The symptom if no hub is connected to the project is that no synchronized table will update or open, hence in particular file listings don't update and new files can't be opened. 

I can probably figure out how this could happen in cocalc.com if @haraldschilly or I can hit the bug, and when it happens capture the full log of the hub-websocket server that our browser is connected to, and also the project log.  This is almost certainly a bug in hub-websocket so it's critical to get that log.
